### PR TITLE
optimize/functions: add BraninHoo test function

### PR DESCRIPTION
**Pull request by btracey**

### CODE REVIEW

**Comment by btracey** [comment]
PTAL @vladimir-ch

**Comment by btracey** [comment]
Updates #63

File: optimize/functions/functions.go [hunk 1]
```diff
@@ -482,6 +482,46 @@ func (Box3D) Minima() []Minimum {
 	}
 }
 
+// BraninHoo implements the Branin-Hoo function. BraninHoo is a 2-dimensional
+// test function for global optimization routines. It has three global minima
```
Comment (vladimir-ch) [inline optimize/functions/functions.go]: Why for global optimization? Whatever even a local method finds will be a global minimum. Maybe "for gradient-free optimization methods" although the function is smooth and `Grad` could be easily added.

File: optimize/functions/functions.go [hunk 2]
```diff
@@ -482,6 +482,46 @@ func (Box3D) Minima() []Minimum {
...
+// with the same function value. The function is
+//  f(x) = (x_2 - (5.1/(4π^2))*x_1^2 + (5/π)*x_1 - 6)^2 + 10*(1-1/(8π))cos(x_1) + 10
+// It has a minimum value of 0.397887 at x^* = {(-π, 12.275), (π, 2.275), (9.424, 2.475)}
```
Comment (vladimir-ch) [inline optimize/functions/functions.go]: Please use 9.424778 for the first component of the last minimum, it gives a more accurate function value.

File: optimize/functions/functions.go [hunk 3]
```diff
@@ -482,6 +482,46 @@ func (Box3D) Minima() []Minimum {
...
+//  https://www.sfu.ca/~ssurjano/branin.html
+type BraninHoo struct{}
+
+func (BraninHoo) Func(x []float64) float64 {
+	if len(x) != 2 {
+		panic("functions: dimension of the problem must be 2")
+	}
+	a, b, c, r, s, t := 1.0, 5.1/(4*math.Pi*math.Pi), 5/math.Pi, 6.0, 10.0, 1/(8*math.Pi)
+
+	term := (x[1] - b*x[0]*x[0] + c*x[0] - r)
```
Comment (vladimir-ch) [inline optimize/functions/functions.go]: Without `()`

File: optimize/functions/functions.go [hunk 4]
```diff
@@ -482,6 +482,46 @@ func (Box3D) Minima() []Minimum {
...
+			F:      0.397887,
+			Global: true,
+		},
+		{
+			X:      []float64{math.Pi, 2.275},
+			F:      0.397887,
+			Global: true,
+		},
+		{
+			X:      []float64{9.424, 2.475},
```
Comment (vladimir-ch) [inline optimize/functions/functions.go]: 9.424778 also here. This point does not give the listed function value.

File: optimize/functions/functions.go (hunk 2 above)
Comment (btracey) [inline optimize/functions/functions.go]: Done.

File: optimize/functions/functions.go (hunk 3 above)
Comment (btracey) [inline optimize/functions/functions.go]: Done.

File: optimize/functions/functions.go (hunk 4 above)
Comment (btracey) [inline optimize/functions/functions.go]: Done.

**Comment by btracey** [comment]
I'm working on a PR that implements the "many local minima" functions. I updated the documentation style here to more closely match that.

**Pull request merged**